src/scanner.c

```diff
@@ -0,0 +1,442 @@
+#include <assert.h>
+#include <stdio.h>
+#include <string.h>
+#include <tree_sitter/parser.h>
+#include <wctype.h>
+
+/**
+ * Debugging helper macros. Example output:
+ *
+ * > HEREDOC_START
+ * scan_start() <-
+ * next E
+ * next O
+ * next F
+ * del EOF
+ * set HEREDOC_START
+ * stop \n
+ * next \n
+ * scan_delimiter() <-
+ * scan_delimiter() -> false
+ * scan_start() -> true
+ *
+ * > HEREDOC_START_NEWLINE HEREDOC_BODY HEREDOC_END_NEWLINE HEREDOC_END
+ * scan_body() <-
+ * next \n
+ * stop $
+ * scan_delimiter() <-
+ * scan_delimiter() -> false
+ * set HEREDOC_START_NEWLINE
+ * scan_body() -> true
+ */
+
+#define debug 0
+
+#define print(...) \
+  if (debug) printf(__VA_ARGS__)
+
+#define peek() lexer->lookahead
+
+#define next()                       \
+  {                                  \
+    print("next %s\n", str(peek())); \
+    lexer->advance(lexer, false);    \
+  }
+
+#define skip()                       \
+  {                                  \
+    print("skip %s\n", str(peek())); \
+    lexer->advance(lexer, true);     \
+  }
+
+#define stop()                       \
+  {                                  \
+    print("stop %s\n", str(peek())); \
+    lexer->mark_end(lexer);          \
+  }
+
+#define set(symbol)                        \
+  {                                        \
+    print("set %s\n", TokenTypes[symbol]); \
+    lexer->result_symbol = symbol;         \
+  }
+
+#define ret(function, result)                                   \
+  print("%s() -> %s\n", function, (result) ? "true" : "false"); \
```

Comment:
why not use `__func__` instead of passing in `function`

Reply:
I'm not sure, I just copied these debug macros over from the c++ scanner

Reply:
no worries, we shouldn't care about that in this PR then. I created #55 to track.